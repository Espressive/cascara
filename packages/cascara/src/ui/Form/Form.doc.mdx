---
title: Form
# propTable: Form.js
status: 'Draft'
---
