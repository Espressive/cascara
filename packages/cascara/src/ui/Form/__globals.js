const WARNING_STRINGS = {
  MISSING_STATE_FORM: 'Form requires dataDisplay to present a form component.',
};

export { WARNING_STRINGS };
