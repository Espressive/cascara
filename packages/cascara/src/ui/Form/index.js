export { default, propTypes as formPropTypes } from './Form';
