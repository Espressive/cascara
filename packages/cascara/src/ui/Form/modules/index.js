// Form-specific Action Modules
import ActionEdit from './ActionEdit';
import FormRow from './FormRow';

export const formActionModules = {
  edit: ActionEdit,
};

export const formModules = {
  row: FormRow,
};
