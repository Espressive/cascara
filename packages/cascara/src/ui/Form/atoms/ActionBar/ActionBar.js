import React from 'react';
import pt from 'prop-types';

// external prop types
import { propTypes as actionButtonPT } from '../../../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

import styles from '../../Form.module.scss';
import ErrorBoundary from '../../../../shared/ErrorBoundary';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
};

const ActionBar = ({ actions }) => {
  return (
    <ErrorBoundary>
      <div className={styles.FormActionBar}>{actions}</div>
    </ErrorBoundary>
  );
};

ActionBar.propTypes = propTypes;

export default ActionBar;
