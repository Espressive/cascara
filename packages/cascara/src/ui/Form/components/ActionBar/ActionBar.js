import React from 'react';
import pt from 'prop-types';

// external prop types
import { propTypes as actionButtonPT } from '../../../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

import styles from '../../Form.module.scss';
import { Boundaries } from '../../../../system-components';

import classnames from 'classnames/bind';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
  className: pt.string,
};

const cx = classnames.bind(styles);

const ActionBar = ({ actions, className }) => {
  return (
    <Boundaries>
      <div className={cx(className, 'FormActionBar')}>{actions}</div>
    </Boundaries>
  );
};

ActionBar.propTypes = propTypes;

export default ActionBar;
