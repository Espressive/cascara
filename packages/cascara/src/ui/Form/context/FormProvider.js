import React from 'react';
import { ModuleContext, ModuleProvider } from '../../../modules/context';
import { useForm } from 'react-hook-form';

const FormProvider = ({ children, value, ...props }) => {
  const formMethods = useForm();
  const { handleSubmit } = formMethods;

  const onSubmit = (data) => {
    // eslint-disable-next-line no-console
    console.table(data);
  };

  const mergedValues = {
    ...ModuleContext.defaultValue,
    ...value,
    formMethods,
  };

  return (
    <ModuleProvider value={mergedValues} {...props}>
      <form onSubmit={handleSubmit(onSubmit)}>{children}</form>
    </ModuleProvider>
  );
};

export default FormProvider;
