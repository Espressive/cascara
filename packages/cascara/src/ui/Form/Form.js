import React, { useState } from 'react';
import pt from 'prop-types';
// import styles from './Form.module.scss';

import ErrorBoundary from '../../shared/ErrorBoundary';
import FormProvider from './context/FormProvider';
import ModuleError from '../../modules/ModuleError';

import { actionModules, dataModules } from '../../modules/ModuleKeys';
import { formActionModules, formModules } from './modules';
import ActionBar from './atoms/ActionBar';

// there are two types of actions a form supports:
//
// a - actions modules compatible with form
// b - form-specific modules like submit, reset, clear
//
// availableActionModules group them together
const availableActionModules = {
  ...actionModules,
  ...formActionModules,
};
const formDataModules = { ...formModules, ...dataModules };
const actionModuleOptions = Object.keys(availableActionModules);
const dataModuleOptions = Object.keys(formDataModules);

const propTypes = {
  // Actions will be appended to each row, they'll appear as buttons.
  actions: pt.shape({
    modules: pt.arrayOf(
      pt.shape({
        module: pt.oneOf(actionModuleOptions).isRequired,
      })
    ),

    // Resolve allowed actions.
    // A function that returns the actions available to the current state of the form
    resolveAllowedActions: pt.func,
  }),

  // An object of modules to display.
  //
  // Every parameter in this object can potentially be rendered in the form.
  data: pt.shape({}),

  // Here you can describe each of the visible columns in your table.
  // @brian, should we call this `fields` istead?
  dataDisplay: pt.arrayOf(
    pt.shape({
      module: pt.oneOf(dataModuleOptions).isRequired,
    })
  ),

  // a form can be editable
  isEditable: pt.bool,

  // A form can start in an editing state
  isInitialEditing: pt.bool,

  // A form can emit events on every action
  onAction: pt.func,
};

// separates the props into two groups:
//
// - myProps, all props defined in `propTypes`
// - otherProps, the rest of the props
//
// @param {Object} props, the props object to filter
// @param {Object} propTypes, a prop-types definition object
// @returns {object}
const filterProps = (props, propTypes) => {
  const myPropKeys = Object.keys(propTypes);

  return Object.entries(props).reduce(
    (filteredProps, [propName, propValue]) => {
      if (myPropKeys.includes(propName)) {
        filteredProps.myProps[propName] = propValue;
      } else {
        filteredProps.otherProps[propName] = propValue;
      }

      return filteredProps;
    },
    {
      myProps: {},
      otherProps: {},
    }
  );
};

const formFields = (display, data) => {
  const renderField = (field) => {
    const { module, label, ...rest } = field;
    const Module = dataModules[module];
    const moduleValue = data[field.attribute];
    const key = `${module}.${field.attribute}.${moduleValue}`;

    // eslint-disable-next-line react/forbid-foreign-prop-types -- @brian we need to see if this approach is what we want
    const { myProps: moduleProps } = filterProps(rest, Module.propTypes);

    return Module ? (
      <Module {...moduleProps} key={key} label={label} value={moduleValue} />
    ) : (
      <ModuleError moduleName={module} moduleOptions={dataModuleOptions} />
    );
  };

  const renderFields = (fields) => fields.map((field) => renderField(field));

  return display?.map((field, i) => {
    const { attribute, module, fields = [] } = field;

    // Check to see if we have a form module, which will probably only be a FormRow
    const FormModule = formModules[module];

    const key = fields.reduce(
      (key, { attribute }) => `${key}.${i}.${attribute}`,
      `${attribute ? `field.${module}.${attribute}` : 'row'}`
    );

    return (
      <ErrorBoundary key={key}>
        {FormModule ? (
          <FormModule {...field}>{renderFields(fields)}</FormModule>
        ) : (
          renderField(field)
        )}
      </ErrorBoundary>
    );
  });
};

const renderActions = (actions) =>
  actions?.modules?.map((action, idx) => {
    const { module, ...rest } = action;
    const Action = availableActionModules[module];

    // In certain predefined-action modules in which a label is not required, e.g. `edit`,
    // the following unique key generation fails, as it relies on the label.
    const key = `${idx}.${module}.${rest.label || module}`;

    // TODO: create a function to filter out props that we don't want
    // eslint-disable-next-line react/forbid-foreign-prop-types -- @brian we need to see if this approach is what we want
    const { myProps } = filterProps(rest, Action.propTypes);

    return Action ? (
      <Action key={key} {...myProps} />
    ) : (
      <ModuleError
        key={key}
        moduleName={module}
        moduleOptions={actionModuleOptions}
      />
    );
  });

const Form = ({
  actions,
  data,
  dataDisplay,
  onAction = () => {},
  isEditable: incomingIsEditable,
  isInitialEditing = false,
  ...rest
}) => {
  const renderedActions = renderActions(actions);
  const isEditable =
    typeof incomingIsEditable === 'undefined'
      ? Boolean(actions?.modules?.find((action) => action.module === 'edit'))
      : incomingIsEditable;

  const [isEditing, setIsEtiding] = useState(() =>
    !isEditable ? false : isInitialEditing
  );

  const fields = formFields(dataDisplay, data);
  function enterEditMode(recordId) {
    setIsEtiding(true);
  }

  function exitEditMode() {
    setIsEtiding(false);
  }

  return (
    <ErrorBoundary>
      <FormProvider
        value={{
          data,
          enterEditMode,
          exitEditMode,
          isEditable,
          isEditing,
          onAction,
        }}
        {...rest}
      >
        {fields}
        <ActionBar actions={renderedActions} />
      </FormProvider>
    </ErrorBoundary>
  );
};

Form.propTypes = propTypes;
Form.displayName = 'Form';

export { propTypes };
export default Form;
