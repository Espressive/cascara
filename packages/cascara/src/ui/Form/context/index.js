import FormProvider from './FormProvider';

export { FormProvider };
