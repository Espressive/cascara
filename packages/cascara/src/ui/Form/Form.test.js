import React from 'react';
import faker from 'faker';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'mutationobserver-shim';

import Form from './';

faker.seed(1);

describe('Form', () => {
  const safeDate = new Date('2021-11-02T22:22:33.405Z');
  const data = {
    avatar: faker.image.avatar(),
    country: faker.address.countryCode(),
    date: safeDate,
    department: faker.name.jobArea(),
    eid: faker.random.uuid(),
    firstName: faker.name.firstName(),
    homePhone: faker.phone.phoneNumber(),
    lastName: faker.name.lastName(),
    officePhone: faker.phone.phoneNumber(),
    title: faker.name.jobTitle(),
  };

  const dataConfig = {
    actions: [
      {
        cancelLabel: 'Cancelar',
        dataTestIDs: {
          cancel: 'cancelButton',
          edit: 'editButton',
          save: 'saveButton',
        },
        editLabel: 'Editar',
        module: 'edit',
        saveLabel: 'Guardar',
      },
      {
        actionName: 'delete',
        'data-testid': 'deleteButton',
        isLabeled: false,
        label: 'Delete',
        module: 'button',
        size: 'small',
      },
    ],
    display: [
      {
        attribute: 'eid',
        label: 'EID',
        module: 'text',
      },
      {
        fields: [
          {
            attribute: 'firstName',
            label: 'First Name',
            module: 'text',
          },
          {
            attribute: 'lastName',
            label: 'Last Name',
            module: 'text',
          },
          {
            attribute: 'officePhone',
            label: 'Office Phone',
            module: 'text',
          },
        ],
        module: 'row',
        ratio: [1, 1, 2],
      },
      {
        attribute: 'homePhone',
        label: 'Home Phone',
        module: 'text',
      },
      {
        fields: [
          {
            attribute: 'title',
            isEditable: false,
            label: 'Title',
            module: 'text',
          },
          {
            attribute: 'department',
            isEditable: false,
            label: 'Department',
            module: 'text',
          },
          {
            attribute: 'country',
            isEditable: false,
            label: 'Country',
            module: 'text',
          },
        ],
        module: 'row',
      },
    ],
  };

  let view;

  beforeAll(() => {});

  test('snapshot test', () => {
    view = render(<Form data={data} dataConfig={dataConfig} />).container;

    expect(view).toMatchSnapshot();
  });

  test('component tree', async () => {
    render(<Form data={data} dataConfig={dataConfig} />);

    const eid = await screen.findByText('EID');
    const firstName = await screen.findByText('First Name');
    const lastName = await screen.findByText('Last Name');
    const homePhone = await screen.findByText('Home Phone');
    const officePhone = await screen.findByText('Office Phone');
    const title = await screen.findByText('Title');
    const department = await screen.findByText('Department');
    const country = await screen.findByText('Country');

    const editButton = await screen.findByTestId('editButton');
    const deleteButton = await screen.findByTestId('deleteButton');

    expect(eid).toBeInTheDocument();
    expect(firstName).toBeInTheDocument();
    expect(lastName).toBeInTheDocument();
    expect(homePhone).toBeInTheDocument();
    expect(officePhone).toBeInTheDocument();
    expect(title).toBeInTheDocument();
    expect(department).toBeInTheDocument();
    expect(country).toBeInTheDocument();

    expect(editButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();
  });

  test('it starts in edit mode when isInitialEditing is true', async () => {
    render(<Form data={data} dataConfig={dataConfig} isInitialEditing />);

    const cancelButton = await screen.findByTestId('cancelButton');
    const saveButton = await screen.findByTestId('saveButton');

    expect(cancelButton).toBeInTheDocument();
    expect(saveButton).toBeInTheDocument();

    const editButton = screen.queryByTestId('editButton');
    expect(editButton).not.toBeInTheDocument();
  });

  test('general flow: View Mode -> Edit Mode -> Save Action -> View Mode', async () => {
    const onAction = jest.fn();

    render(<Form data={data} dataConfig={dataConfig} onAction={onAction} />);

    // make sure we are in view mode
    let cancelButton = screen.queryByTestId('cancelButton');
    let saveButton = screen.queryByTestId('saveButton');
    let editButton = screen.queryByTestId('editButton');

    expect(cancelButton).not.toBeInTheDocument();
    expect(saveButton).not.toBeInTheDocument();
    expect(editButton).toBeInTheDocument();

    // enter edit mode
    userEvent.click(editButton);

    // make sure we are in edit mode
    cancelButton = await screen.findByTestId('cancelButton');
    saveButton = await screen.findByTestId('saveButton');
    editButton = screen.queryByTestId('editButton');

    expect(cancelButton).toBeInTheDocument();
    expect(saveButton).toBeInTheDocument();
    expect(editButton).not.toBeInTheDocument();

    // save button should be disabled when form is pristine
    expect(saveButton).toBeDisabled();

    // edit some data
    let firstName = await screen.findByLabelText('First Name');
    expect(firstName).toBeInTheDocument();

    userEvent.type(firstName, 'nio');

    saveButton = await screen.findByTestId('saveButton');
    expect(saveButton).not.toBeDisabled();

    // Save action
    userEvent.click(saveButton);

    // make sure we are back to view mode
    editButton = await screen.findByTestId('editButton');
    cancelButton = screen.queryByTestId('cancelButton');
    saveButton = screen.queryByTestId('saveButton');

    expect(cancelButton).not.toBeInTheDocument();
    expect(saveButton).not.toBeInTheDocument();
    expect(editButton).toBeInTheDocument();

    // check that onAction event was triggered
    expect(onAction).toHaveBeenCalledTimes(2);

    expect(onAction).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'edit.start',
      }),
      expect.objectContaining(data)
    );

    // check that the data is actually updated in the event
    expect(onAction).toHaveBeenLastCalledWith(
      expect.objectContaining({
        name: 'edit.save',
      }),
      expect.objectContaining({
        ...data,
        firstName: `${data.firstName}nio`,
      })
    );
  });

  test('onAction events', async () => {
    const onAction = jest.fn();

    render(<Form data={data} dataConfig={dataConfig} onAction={onAction} />);

    // verify buttons are there
    let deleteButton = screen.queryByTestId('deleteButton');
    let editButton = screen.queryByTestId('editButton');

    expect(deleteButton).toBeInTheDocument();
    expect(editButton).toBeInTheDocument();

    // enter edit mode
    userEvent.click(deleteButton);

    // check that onAction event was triggered
    expect(onAction).toHaveBeenCalled();
  });
});
