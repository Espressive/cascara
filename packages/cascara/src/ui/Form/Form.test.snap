// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshot test 1`] = `
<div>
  <form
    class="Form"
  >
    <div
      class="Text"
    >
      <span>
        <span
          class="LabelText"
        >
          EID
        </span>
        <span
          class="Input"
        >
          e024f231-6265-4a6e-8d65-a837e308ae67
        </span>
      </span>
    </div>
    <div
      class="FormRow"
      style="grid-template-columns: 1fr 1fr 2fr;"
    >
      <div
        class="Text"
      >
        <span>
          <span
            class="LabelText"
          >
            First Name
          </span>
          <span
            class="Input"
          >
            Keaton
          </span>
        </span>
      </div>
      <div
        class="Text"
      >
        <span>
          <span
            class="LabelText"
          >
            Last Name
          </span>
          <span
            class="Input"
          >
            Streich
          </span>
        </span>
      </div>
      <div
        class="Text"
      >
        <span>
          <span
            class="LabelText"
          >
            Office Phone
          </span>
          <span
            class="Input"
          >
            (808) 221-0860 x54694
          </span>
        </span>
      </div>
    </div>
    <div
      class="Text"
    >
      <span>
        <span
          class="LabelText"
        >
          Home Phone
        </span>
        <span
          class="Input"
        >
          391-887-9830 x658
        </span>
      </span>
    </div>
    <div
      class="FormRow"
      style="grid-template-columns: repeat(auto-fill, minmax(10em, 1fr));"
    >
      <div
        class="Text"
      >
        <span>
          <span
            class="LabelText"
          >
            Title
          </span>
          <span
            class="Input"
          >
            Global Implementation Developer
          </span>
        </span>
      </div>
      <div
        class="Text"
      >
        <span>
          <span
            class="LabelText"
          >
            Department
          </span>
          <span
            class="Input"
          >
            Assurance
          </span>
        </span>
      </div>
      <div
        class="Text"
      >
        <span>
          <span
            class="LabelText"
          >
            Country
          </span>
          <span
            class="Input"
          >
            ZW
          </span>
        </span>
      </div>
    </div>
    <div
      class="FormActionBar"
    >
      <button
        class="ui basic button"
        data-testid="editButton"
        type="button"
      >
        Editar
      </button>
      <button
        class="ui basic button"
        data-testid="deleteButton"
        name="delete"
        type="button"
      >
        ActionButton
      </button>
    </div>
  </form>
</div>
`;
